File: my-folder/0015-3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        int n=nums.size();
        if(n==0 or n==1 or n==2)
            return res;
        sort(nums.begin(),nums.end());
        int l=0,r=0,sum=0,a=0;
        
        for(int i=0;i<n-2;)
        {
            l=i+1;
            r=n-1;
            a=-nums[i];
            
            while(l<r)
            {
                sum=nums[l]+nums[r];
                if(sum<a)
                {
                    while(++l<n and nums[l]==nums[l-1]);
                }
                else if(sum>a)
                {
                    while(--r>l and nums[r]==nums[r+1]);
                }
                else
                {
                    res.push_back({nums[i],nums[l],nums[r]});
                    while(++l<n and nums[l]==nums[l-1]);
                    while(--r>l and nums[r]==nums[r+1]);  
                }
            }
            while(++i<n and nums[i]==nums[i-1]);
        }
        return res;
        
    }
};

File: my-folder/0015-3sum/solution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);
        int length = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < length; i++) {
            if (nums[i] > 0)
                break;
            if (i > 0 && nums[i - 1] == nums[i])
                continue;

            int l = i + 1, r = nums.length - 1;

            while (l < r) {
                int sum = nums[l] + nums[r] + nums[i];
                if (sum == 0) {
                    LinkedList<Integer> temp = new LinkedList<>(Arrays.asList(nums[i], nums[l], nums[r]));
                    res.add(temp);
                    l++;
                    r--;
                    while (l < r && nums[l - 1] == nums[l]) {
                        l++;
                    }
                } else if (sum > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }

        return res;

    }

}
